core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -468,21 +480,46 @@ private void handleCacheStreamException(IOException e) {
   }
 
   /**
-   * Only updates {@link #mCurrentCacheStream}, {@link #mCurrentBlockInStream} and
-   * {@link #mStreamBlockId} to be in sync with the current block (i.e. {@link #getCurrentBlockId()}
-   * ). If this method is called multiple times, the subsequent invokes become no-op. Call this
-   * function every read and seek unless you are sure about the block streams are up-to-date.
+   * Checks whether the current block stream has remaining bytes, if not, advance to the next block
+   * stream, otherwise, this is a no-op. This should only be used in streaming read APIs.
+   * For APIs like {@link #seek(long)}, {@link #skip(long)}, the block stream should be updated
+   * according to the current read position, use {@link #updateStreamsBasedOnPosition()} instead.
    */
   private void updateStreams() throws IOException {
+    assureCacheStreamInSync();
+    if (shouldUpdateStreams()) {
+      mBlockIdIndex++;
+      updateStreamsInternal();
+    }
+  }
+
+  /**
+   * Computes the current block stream ID based on the current read position, if the stream ID is
+   * different from the current block stream ID, then updates the stream.
+   * Comparing to {@link #updateStreams()}, this method has overhead for computing the
+   * block stream ID, only use this method for non-streaming read APIs like {@link #seek(long)}
+   * and {@link #skip(long)}.
+   */
+  private void updateStreamsBasedOnPosition() throws IOException {
+    assureCacheStreamInSync();
+    int blockIdIndex = (int) (mPos / mBlockSize);
```

Comment:
this is only for check, right? can we simply do:
```java
long blockStartPos = mBlockSize * mBlockIdIndex;
long blockEndPos = blockStart + mBlockSize;
if (mPos < blockStartPos || mPos >= blockEndPos) {
  ...
}
```

The reason is multiplication is 5 to 10 times faster in general than division

Reply:
nop, the new block index has to be computed based on the position, `mPos` might have been changed in a `seek` or `skip`, this method is to update the streams according to the recomputed block index.